Config Changes and New 'replace' Feature

bevhost opened the pull request:
First of all I found that the postfix was not considered in the comparison therefore all prefixes had to have a unique forward or prefix. 
Another way to solve this problem is to encode then entire IP Address, rather than the offset within the range.
This is typical of many existing ip4 implementations which use reverse entries like this
cpe-1-1-168-192.rev.example.net
To achieve this I added a new config keyword "replace" that allows you to specify a character which can be used to replace the . in IP4 or the : in IP6 addresses. e.g. u1--00fd-l0.example.com

When the entire IP is encoded prefix & postfix become optional and forward domains do not need to be unique.

To ensure that the new changes continue to work on old configs, there is an extra piece of code to make replace default to None.  I expect that where replace is used, it would be set to "-". 

Another thing I discovered is that the IP version doesn't need to specified in the configuration as it can be calculated.

I've also updated to test files to work with the changes made in the previous PR.
e.g.
```
defaults:
  email : 'dave.bevhost.net'
  dns : 'dns.bevhost.net'
  ttl : 900
  prefix: ''
  replace: '-'
  postfix: '-ip4'
  forward: 'bevhost.com'
  nameserver:
    - dns.bevhost.net
    - dns.bevhost.com
prefixes:
  '2401:d002:4a02:200::/56':
    postfix: '-ip6'
    replace:
  '125.253.8.16/28':
  '10.2.47.0/24':
  '192.168.139.0/24':
```

You can see how I'm using the replace feature on all IP4 blocks so that I can list them all with no other parameters, relying solely on the defaults.  This becomes more convenient the more blocks you have,

bevhost committed:
Match Postfix as well
required to tell them apart when ranges have same prefix and forward domain

bevhost committed:
Calculate version so it's no longer needed in config

bevhost committed:
Remove blank line

bevhost committed:
Change test files to work with new base32

bevhost committed:
Add replace feature
allows 192-168-0-0.example.net format

bevhost committed:
Version is now calculated automatically

Commit:
add utility to list reverse zones

Commit:
die after a while and let pdns restart us
avoids issue where process hangs eventually
add suggestion about pipe-regex

bevhost committed:
Check that IP address is within range

bevhost committed:
Revert "Check that IP address is within range"

This reverts commit 7ed7fdea299ebd6930d61cb0d8f7d22dc48cf44e.

bevhost committed:
Check that IP address is within range

Commit:
Validate String

Commit:
Add Data Validation

endreszabo commented:
geez, I should have made this repo an archive years ago :)

anyway, thanks for the PR and I'll review it shortly.

Commit:
Use pathfix.py to set the shebang to python version

Commit:
use safe_load

bevhost committed:
- Remove pathfix.py, hardcode to python 3, so python 2.7 no longer supported
- Replace IPy module with Python 3 native ipaddress module (vibe code)
- Create get_reverse_pointer() and get_reverse_names() helper functions
- Update parse_config() to use ipaddress.ip_network()
- Remove IPy from requirements.txt and Requires: since ipaddress is built-in

bevhost committed:
Use safe_load for show-reverse-zones.p as well

bevhost committed:
Add selinux files

endreszabo commented:
The recent IPy → ipaddress migration broke get_reverse_pointer() — it returns the host-level reverse pointer (0.0.254.169.in-addr.arpa) instead of the zone-level one (254.169.in-addr.arpa).

bevhost committed:
Get reverse pointer for an IP network zone (not host-level).

bevhost commented:
Thank-you for reporting that.  I think I have fixed it. However,
I have just become aware of one more issue during testing for that.  
It's returning hostnames that are out of bounds. e.g. for 192.168.0.0/16 
255.255.168.192.in-addr.arpa domain name pointer 2661-ip4.bevhost.net.
2661-ip4.bevhost.net has address 192.168.255.255
2667-ip4.bevhost.net has address 192.169.0.5
I'll look into this as well.

bevhost commented:
FWIW, this website https://www.potaroo.net/ietf/all-ids/draft-ietf-dnsop-isp-ip6rdns-02.html#rfc.section.2.5
States that dynamically generating reverse DNS for IPv6 might be a bad idea.
In practice we have found that the performance of the python script is not as fast as entries returned from a database.
It's possible that too many requests could overwhelm a server.
We generally run multiple servers with AnyCast and always have a fallback server that is missing the python module to handle these situations

bevhost committed:
Check that result is within the range

endreszabo commented:
> Thank-you for reporting that. I think I have fixed it.

Yeah, it was also surprise to me. For the record: I let Claude port your PR to a golang CoreDNS plugin and it was Claude noticing this 'bug'. Never thought of a reply like that.

> 2667-ip4.bevhost.net has address 192.169.0.5
> I'll look into this as well.

We may call this an out-of-range bug, but realistically it would never appear as PTR queries for non-delegated IP ranges would not hit the nameservers whatsoever.

> States that dynamically generating reverse DNS for IPv6 might be a bad idea.

Well, yes and no, depends on. The world seems to be moving towards forward-based authentication schemes where reverse records don't really matter anymore. Think of DANE, SPF, DKIM, COA and friends. These days it's basically SMTP and IRC left that rely on reverse records. I personally find dynamic reverse records generated very sexy and always shows a professionalism of a given ISP when I see one. Others are not that sophisticated and will have a wildcard record to serve a generic PTR reply or does not even bother and will return NXDOMAIN for general PTR queries.

> It's possible that too many requests could overwhelm a server.

Yes it does have an extra CPU burden for sure, especially when it comes to dnssec sign dynamic records on-the-fly. You can mitigate this kind of being DoSsed this way by having extra level of discrete nameservers serving this dynamic traffic to shave off that extra load from the nameservers serving critical zones.

There are ISPs that do not delegate this to a sub-zone for IP their pools and have the dynamic PTR records under their main brand label zone. That's equally elegant and hazardous: we had a case in Hungary where a user could acquire a let's encrypt certificate for his website serving a domain that was under this ISPs main brand domain:

1.0.0.0.0.0.1.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.2.0.0.1.1.1.0.0.a.2.ip6.arpa. 86400	IN PTR 2A001110020000000000000001000001.mobile.pool.telekom.hu.

With this, the user *could* collect a badly scoped session cookie from the original telekom.hu domain to impersonate legit customers.

Personally I would always have these records under a domain that hints that this is a consumer/partner originated content. This is what github and google do with their githubusercontent.com and such domains. This is also what I do with WireDNS, a WireGuard-based dynamic DNS service where the users' records are assigned from the wirednspeer.net domain and not any of the main wiredns service domains.

> In practice we have found that the performance of the python script is not as fast as entries returned from a database.

By using a systems programming language (like, C, Go, Rust) generating records should be much faster than querying a database for records. I'm 100% sure in that. That's why I moved away from python: 5 ms PowerDNS+Python pipe backend replies turned into sub-millisecond replies with CoreDNS. Once I had a badly configured record generator for MX records yielding into recursive lookups and had a lot of DDoS like feeling traffic hitting my nameservers and that was the time when I looked for other, more performant alternatives. Interestingly, it was only recursive resolvers coming from AWS ranged that did this crazy recursion, even well after I fixed the problem.

> We generally run multiple servers with AnyCast and always have a fallback server that is missing the python module to handle these situations

That's an interesting engineering decision :) I recommend to either delegate zones with dynamic records to a 2nd set of nameservers (=zones with dynamic records delegated) running this python backend or rewriting the stuff in a more performant programming language. As stated, I'm already in the process of porting this to be a CoreDNS plugin and would be happy to help you migrating.
